torch/autograd/_functions/tensor.py

```diff
@@ -601,13 +609,93 @@ def forward(self, input):
+        which works with all types of inputs is slow, namely
+        O(n^2) where n = input.size(self.dim).
+
+        The second one is much faster, namely O(n), but
+        requires nonzero inputs. This implementation checks
+        for 0s in the inputs and then decides which one to use.
+        '''
+        input, = self.saved_tensors
+        dim_size = input.size(self.dim)
+        
+        if dim_size == 1:
+            return grad_output.clone()
+
+        if input.is_cuda:
+            LT = torch.cuda.LongTensor
+        else:
+            LT = torch.LongTensor
+
+        if (input == 0).any():
+            ones = input.index_select(self.dim,
+                    LT(range(1))).fill_(1).clone()
+            zeros = ones.clone().fill_(0)
+            grad_input = grad_output.new(input.size()).zero_() 
+            for k in range(dim_size):
+                if 0 < k and k < dim_size - 1:
+                    prods_until_k = torch.prod(input.index_select(self.dim,
+                        LT(range(k))), dim=self.dim)
+
+                    prods_from_k_plus_1 = torch.cumprod(input.index_select(
+                        self.dim, LT(range(k+1, dim_size))), dim=self.dim)
+
+                    ommitted_products = prods_until_k.expand_as(
+                        prods_from_k_plus_1) * prods_from_k_plus_1
+
+                    ommitted_products = torch.cat((prods_until_k,
+                        ommitted_products), self.dim)
+                elif k == dim_size - 1:
+                    prods_until_k = torch.prod(input.index_select(self.dim,
+                        LT(range(k))), dim=self.dim)
+
+                    ommitted_products = prods_until_k
+                else: #k == 0
+                    prods_from_k_plus_1 = torch.cumprod(input.index_select(
+                        self.dim, LT(range(k+1, dim_size))), dim=self.dim)
+
+                    ommitted_products = torch.cat((ones, prods_from_k_plus_1),
+                        self.dim)
+                
+                # At this point ommitted_products is the same size
+                # as input, except on the dimension dim where it's
+                # dim_size - k
+                assert ommitted_products.size()[self.dim] == dim_size - k, \
+                    "Dimension error"
+
+                if k != 0:
+                    size_to_expand = [l for l in input.size()]
+                    size_to_expand[self.dim] = k # Adding zeros to missing dimensions
+                    size_to_expand = torch.Size(size_to_expand)
+                    expanded_zeros = zeros.expand(size_to_expand)
+                    ommitted_products = torch.cat((zeros.expand(size_to_expand),
+                            ommitted_products), self.dim)
+
+                grad_input.index_copy_(self.dim, LT(range(k,k+1)), 
+                        torch.sum(grad_output * ommitted_products, dim=self.dim))
 
+        else:
+            output = torch.cumprod(input, dim=self.dim)
```

Comment:
I thought about this, but the computational cost of recalculating it seems small. The added cost is doing one cumprod, when we are already making 5 or 6 other operations with the same cost, including one cumsum. Right now we pay a bit more of overhead but store less things for the backward, saving half the memory.

Thoughts?